Migration of ExpressionSyntaxTreeService to Angular 8

h-ARTS opened the pull request:
## Overview
<!--
READ ME FIRST: Please answer *both* questions below and check off every point from the essential checklist!
-->

~1. This PR fixes or fixes part of #[fill_in_number_here].~
2. This PR does the following: Migration of `ExpressionSyntaxTreeService` to Angular 8

## Essential Checklist

- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## PR Pointers

- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays
- Never force push. If you do, your PR will be closed.

h-ARTS committed:
ExpressionSyntaxTreeService added to UpgradedServices

h-ARTS committed:
Migrate ExpressionSyntaxTreeService to Angular 8

h-ARTS commented:
@bansalnitish @brianrodri Can anyone check this out for review and check Travis?

bansalnitish commented on core/templates/expressions/expression-syntax-tree.service.spec.ts:
```diff
@@ -15,40 +15,30 @@
-          $injector.get('ExpressionSyntaxTreeService');
-    }));
+    beforeEach(() => {
+      expressionSyntaxTreeService = new ExpressionSyntaxTreeService(
```
This is fine but can we please use TestBed to create an instance of this service please? This will make our code consistent and that is a more cleaner way of creating an instance. 

Think what is ExpressionParserService() also has some dependency, we would have so many nested calls.

h-ARTS replied:
Done

bansalnitish commented on core/templates/expressions/expression-syntax-tree.service.ts:
```diff
@@ -16,417 +16,491 @@
+      args: Array<number>,
+      expectedNum: number,
+      expectedMax: number = expectedNum
+  ): void {
```
Can we please move this to above line if it's less than 80 characters?

h-ARTS replied:
`args` and `expectedNum` can be placed in the same line. Sorry that was automatically formatted 😊

bansalnitish commented on core/templates/expressions/expression-syntax-tree.service.ts:
```diff
@@ -16,417 +16,491 @@
+  // throws an error if not. If optional expectedMax is specified, it
+  // verifies the number of args is in [expectedNum, expectedMax] range
+  // inclusive.
+  public verifyNumArgs(
```
Also, shouldn't this be private and not public?

h-ARTS replied:
You are right, this method is used only in this file although it wasn't written as `_verifyNumArgs` as per convention for distinguishing private/public

bansalnitish reviewed (commented):
Again amazing stuff @h-ARTS, thank you!

Took first pass and left some comments for you, PTAL!

h-ARTS committed:
Minor changes:

- privatised method verifyNumArgs
- TestBed used for Service
- Lines coerced for verifyNumArgs

h-ARTS reviewed (commented):
@bansalnitish PTAL

bansalnitish commented on core/templates/expressions/expression-syntax-tree.service.ts:
```diff
@@ -16,417 +16,488 @@
+      parsed: string,
+      envs: Array<object>,
+      func: (parsed: string, envs: Array<any>) => string
+  ) {
```
Can we move this above as well please?

bansalnitish commented on core/templates/expressions/expression-syntax-tree.service.ts:
```diff
@@ -16,417 +16,488 @@
+    throw new ExprWrongArgTypeError(
+      originalValue,
+      typeof originalValue,
+      'Number'
```
Can we move some of the args in the same line here as well?

bansalnitish replied:
Ditto below, I see each param in new line and then closing bracket in next line. Can we try to move as many params in one line and the closing bracket as well in the same ending line?

h-ARTS replied:
@bansalnitish PTAL!

bansalnitish reviewed (commented):
It's close! Some minor nits for you.

h-ARTS committed:
Adjusted arguments in one line after autoformat issue

h-ARTS commented:
@bansalnitish @brianrodri Can anyone restart the 2 Travis failing CI's? The last commit was passing.

brianrodri commented:
Done!

h-ARTS commented:
@bansalnitish @brianrodri Sorry for disturbing again. `RUN_E2E_TESTS_COLLECTIONS=true` doesn't want to be green. It's flaking. Rerun?

bansalnitish commented on core/templates/expressions/expression-syntax-tree.service.ts:
```diff
@@ -16,417 +16,445 @@
-          _verifyArgTypesMatchExpectedType(args, PARAMETER_TYPES.REAL);
-          return PARAMETER_TYPES.UNICODE_STRING;
-        }
+      getType: (args: Array<any>): string => {
```
Is this really any or do we don't know the exact type? Ditto here and below.

h-ARTS replied:
You are right, I just console.log them during the test and noticed the `args` array contains strings. I should've consoled them in the first place. 😓

h-ARTS replied:
@bansalnitish It's more complicated then I thought. Here's what happens:
If I set `system.pow.eval`s args to `Array<string>`
<img width="902" alt="Screenshot 2020-03-24 at 11 58 45" src="https://user-images.githubusercontent.com/11369764/77418194-da33c700-6dc6-11ea-8df0-daa16c97720f.png">
`Math.pow`s first argument will throw this error. If I would set it to `Array<number>` then `this.verifyNumArgs`s first parameter would throw a similar error and because I know from the `console.log` you get strings so I was forced to set to `Array<any>`. 😲I could, however, manage the rest of the operators.

Your thoughts on this?

bansalnitish replied:
For these can we set them to `Array<number|string>` then which is far better than `Array<any>`?

Also, aren't these two different functions one uses `Array<string>` the other uses `Array<number>` (didn't go completely through the codebase, first suggestion would anyway work)?

bansalnitish reviewed (approved):
Thanks @h-ARTS, this is good stuff.

h-ARTS committed:
Placed exact types for evaluation and getter method.

- short-circuit comment removed as it's been already done using logical/ternary operators

bansalnitish commented on core/templates/expressions/expression-syntax-tree.service.ts:
```diff
@@ -390,54 +389,57 @@ export class ExpressionSyntaxTreeService {
     },
     pow: {
-      eval: (args: Array<any>): number => {
+      eval: (args: Array<number|string>): number => {
```
Why we need `string` here? On which line exactly does marking argument as `Array<number>` throw error?

h-ARTS replied:
@bansalnitish It used to throw on `verifyNumArgs` method because I've set the args on that method to `Array<number|string>`. Do we actually have a specification for these operators, if, pow, log and abs? Otherwise, I'll find it hard to figure out what `+` should evaluate. 
So far what I thought of these use cases: `3 + 'text as string'` or `-2` or `'hello world!'` which turns out that I needed to set `args` to `Array<number|string>` same for `pow`. 

Correct me if I'm wrong 😲

h-ARTS replied:
@bansalnitish I've sent you an invitation on hangouts

bansalnitish replied:
1. Correct me if I'm wrong -- `verifyNumArgs` expects args to be `string` and Math.functions expect args to be just `number`. However it should be just number for Math related functions hence you added a + sign.

2. What does _coerceAllArgsToNumber(args) expect args to be?

3. I didn't get any invitation on Hangouts.

h-ARTS replied:
@bansalnitish <img width="395" alt="Screenshot 2020-03-25 at 17 03 13" src="https://user-images.githubusercontent.com/11369764/77557875-8acec300-6eba-11ea-8a8c-5043dc014c4b.png">

h-ARTS replied:
@bansalnitish  
2. `_coerceAllArgsToNumber(args: Array<any>)`
When I set it to `Array<number|string>` or only `Array<string>` I get an error like this because `_coerceToNumber()` originalValue expects to be string.
<img width="1020" alt="Screenshot 2020-03-25 at 18 16 37" src="https://user-images.githubusercontent.com/11369764/77565790-05044500-6ec5-11ea-97ec-cdf46e8455c8.png">

bansalnitish replied:
Okay I understand the issue. Let's mark it Array<number|string> as we did initially and use the + sign that you've done. 

I'm unable to find your invite. I checked my account searching your name and even on hangouts.google.com. Am I missing something?

Showtim3 replied:
@bansalnitish, try clicking this **Invitations**
![image](https://user-images.githubusercontent.com/30312043/77854164-356e1b00-7206-11ea-9325-a5f064221577.png)

bansalnitish commented on core/templates/expressions/expression-syntax-tree.service.ts:
```diff
@@ -16,417 +16,447 @@
+  // Coerces all values in the given argument array to Number, and throws
+  // an error if the result is NaN.
+  private _coerceAllArgsToNumber(args: Array<any>): Array<number> {
+    // NOTE TO DEVELOPERS: Setting Array<number|string> and expecting
```
Let's write this clearly. Something like:
```
Setting args to Array<number|string> throws this error (mention the error) and this._coerceToNumber(args[i]) since _coerceToNumber expect args to be just strings.
```

h-ARTS replied:
Done

Showtim3 replied:
@h-ARTS If you think, that it's done and won't require any further comment for the reviewer you can go ahead and resolve this particular comment.

bansalnitish commented on core/templates/expressions/expression-syntax-tree.service.ts:
```diff
@@ -16,417 +16,447 @@
+      eval: (args: Array<number|string>): number => {
+        this.verifyNumArgs(args, 2);
+        var numericArgs = this._coerceAllArgsToNumber(args);
+        // NOTE TO DEVELOPERS: we need to coerce args items with
```
Please change this comment as well. Clearly mentioning why we need + here. It's okay if the comment becomes too long, it should be clear to anyone who reads it even after years :).

h-ARTS replied:
@bansalnitish I was rewriting the comment until one variable caught my eyes 🤭
`pow: eval: var numericArgs = this._coerceAllArgsToNumber(args)` is indicating me in my editor that it's unused. I mean gosh the whole time I was debugging and I should've seen this in the first place! So embarrassing 😓

Result:
```
eval: (args: Array<string>): number => {
        this.verifyNumArgs(args, 2);
        var numericArgs = this._coerceAllArgsToNumber(args);
        return Math.pow(numericArgs[0], numericArgs[1]);
      },
```

bansalnitish replied:
No worries, I'm glad you got it :)

bansalnitish reviewed (commented):
Left some minor nits for you again :P

h-ARTS committed:
Math.pow.eval number args fix by using the numericArgs variable and type corrections

bansalnitish commented:
@brianrodri could you do the code owner review please?

h-ARTS commented:
@brianrodri PTAL

brianrodri commented on core/templates/expressions/expression-syntax-tree.service.ts:
```diff
@@ -16,417 +16,447 @@
-      return this.name + ': ' + this.varname + ' not found in ' + this.envs;
-    };
+// Exceptions that can be thrown from the evaluation of expressions.
+const ExpressionError = function(): void {};
```
Please make these all classes. Our previous code tried to emulate classes by manually changing prototypes, but now that we are migrating to Angular 8 which uses typescript, that's no longer necessary. As part of the migration, we should be changing these to classes like so:

```typescript
public class ExpressionError {
}

public class ExprUndefinedVarError {
  constructor(public varname: string, public envs: any[]);

  toString(): string {
    // ...
  }
}

// etc...
```

h-ARTS replied:
@brianrodri Are you sure we can declare nested classes? It doesn't work and not even in TS playground.

brianrodri replied:
They shouldn't be nested, they should be top-level.

brianrodri commented on core/templates/expressions/expression-syntax-tree.service.ts:
```diff
@@ -16,417 +16,447 @@
-      }
-      return true;
-    };
+    return _findParams(parsed);
```
Since we're only using this function once, let's just extract the body and get rid of it.

h-ARTS replied:
@brianrodri It's also used when concatenating params recursively or can you help me understand this? Maybe I might be mistaken. 🤭

brianrodri replied:
Ah didn't catch that. Let's extract it out and make it a private method instead, then?

h-ARTS replied:
@brianrodri Alright 👍

brianrodri commented on core/templates/expressions/expression-syntax-tree.service.ts:
```diff
@@ -16,417 +16,447 @@
+    if (expectedMax === undefined) {
+      expectedMax = expectedNum;
+    }
+    if (args.length >= expectedNum && args.length <= expectedMax) {
```
Please invert this condition and throw from it instead.

h-ARTS replied:
@brianrodri If I would invert this condition and throw from it like so
```
if (args.length <= expectedMax && args.length >= expectedNum) {
      throw new ExprWrongNumArgsError(args, expectedNum, expectedMax);
 }
```
4 tests will fail even if I would place on top of the first if statement:
1x `Error: Expected 'abc<oppia-expression-error-tag></oppia-expression-error-tag>' to equal 'abc3'.`
3x `TypeError: Right-hand side of 'instanceof' is not an object`

brianrodri replied:
Careful, the opposite of `a && b` is `!a || !b` (from the [De Morgan's laws](https://en.wikipedia.org/wiki/De_Morgan%27s_laws)), and the opposite of `a <= b` is `a > b` so you should change it to:

```typescript
if (args.length < expectedNum || args.length > expectedMax) {
  throw new ExprWrongNumArgsError(args, expectedNum, expectedMax);
}
```

h-ARTS replied:
@brianrodri I must admit, I just learned something. Thanks!

brianrodri replied:
Happy I could help you learn something new :)

brianrodri commented on core/templates/expressions/expression-syntax-tree.service.ts:
```diff
@@ -16,417 +16,447 @@
+    throw new ExprWrongNumArgsError(args, expectedNum, expectedMax);
+  }
+
+  private _verifyArgTypesMatchExpectedType(argTypes: Array<string>,
```
Please remove leading underscore from all methods. We used it before to emulate "private" functions, but that's no longer necessary now that we're using typescript.

brianrodri commented on core/templates/expressions/expression-syntax-tree.service.ts:
```diff
@@ -16,417 +16,447 @@
-    };
-  }
-]);
+  public ExpressionError = ExpressionError;
```
These should no longer be necessary after declaring them as classes (see my first comment)

h-ARTS replied:
@brianrodri They are used in *expression-interpolation.service.ts* and other services or tests: 
```
if ((e instanceof ExpressionParserService.SyntaxError) ||
   (e instanceof ExpressionSyntaxTreeService.ExpressionError)) {
       return EXPRESSION_ERROR_TAG;
}
```
It won't pass if I would have to remove them unless I could replace it from `ExpressionSyntaxTree.ExpressionError` to only `ExpressionError` and in several other services and tests

brianrodri replied:
Gotcha, in that case it's fine to keep these lines as-is (we should still be able to make them classes)

brianrodri reviewed (changes requested):
Thanks @h-ARTS! There are some more changes we should be making as part of the migration, please see my comments for details.

h-ARTS committed:
underscores removed from all methods

h-ARTS committed:
Expression error types converted to classes

h-ARTS committed:
Extracted recursive findParams private method

h-ARTS committed:
Inverted condition from logical AND to OR

* a >= b && c <= d    to     a < b || c > d

brianrodri commented on core/templates/expressions/expression-syntax-tree.service.ts:
```diff
@@ -23,54 +23,83 @@ import { AppConstants } from 'app.constants';
+    //  we need to adjust the prototype manually right after the super() call
+    //  For more information refer to this link below:
+    //  https://github.com/Microsoft/TypeScript-wiki/blob/master/Breaking-Changes.md#extending-built-ins-like-error-array-and-map-may-no-longer-work
+    Object.setPrototypeOf(this, ExpressionError.prototype);
```
All of the constructors should have an empty body, there isn't anything we need to do in any of them, so please remove this code.

h-ARTS replied:
@brianrodri Sorry if I'm bothering you again but if I would remove this (leaving the constructor body blank except `super()` then tests will fail again. This is why I added this line of code and the error looks like this which I can't follow after investigating it:
```
Expression type parser service should determine the correct types for the expressions FAILED
	Error: Expected true to be false.
```
```
ExprUndefinedVarError thrown
	error properties: Object({ varname: 'a', envs: [ Object({  }), Object({ +: Object({ eval: Function, getType: Function }), -: Object({ eval: Function, getType: Function }), *: Object({ eval: Function, getType: Function }), /: Object({ eval: Function, getType: Function }), %: Object({ eval: Function, getType: Function }), <=: Object({ eval: Function, getType: Function }), >=: Object({ eval: Function, getType: Function }), <: Object({ eval: Function, getType: Function }), >: Object({ eval: Function, getType: Function }), !: Object({ eval: Function, getType: Function }), ==: Object({ eval: Function, getType: Function }), !=: Object({ eval: Function, getType: Function }), &&: Object({ eval: Function, getType: Function }), ||: Object({ eval: Function, getType: Function }), if: Object({ eval: Function, getType: Function }), floor: Object({ eval: Function, getType: Function }), pow: Object({ eval: Function, getType: Function }), log: Object({ eval: Function, getType: Function }), abs: Object({ eval: Function,  ...
	    at <Jasmine>
	    at ExprUndefinedVarError.ExpressionError [as constructor] (core/templates/combined-tests.spec.js:22806:72353)
```

h-ARTS replied:
@brianrodri Sorry if I'm bothering you again but if I would remove this (leaving the constructor body blank except `super()` then tests will fail again. This is why I added this line of code and the error looks like this which I can't follow after investigating it:
```
Expression type parser service should determine the correct types for the expressions FAILED
	Error: Expected true to be false.
```
```
ExprUndefinedVarError thrown
	error properties: Object({ varname: 'a', envs: [ Object({  }), Object({ +: Object({ eval: Function, getType: Function }), -: Object({ eval: Function, getType: Function }), *: Object({ eval: Function, getType: Function }), /: Object({ eval: Function, getType: Function }), %: Object({ eval: Function, getType: Function }), <=: Object({ eval: Function, getType: Function }), >=: Object({ eval: Function, getType: Function }), <: Object({ eval: Function, getType: Function }), >: Object({ eval: Function, getType: Function }), !: Object({ eval: Function, getType: Function }), ==: Object({ eval: Function, getType: Function }), !=: Object({ eval: Function, getType: Function }), &&: Object({ eval: Function, getType: Function }), ||: Object({ eval: Function, getType: Function }), if: Object({ eval: Function, getType: Function }), floor: Object({ eval: Function, getType: Function }), pow: Object({ eval: Function, getType: Function }), log: Object({ eval: Function, getType: Function }), abs: Object({ eval: Function,  ...
	    at <Jasmine>
	    at ExprUndefinedVarError.ExpressionError [as constructor] (core/templates/combined-tests.spec.js:22806:72353)
```

brianrodri replied:
Remove `super()` as well, each of these constructor bodies should literally be: `{}` (not sure why you added `super()` in the first place?)

h-ARTS replied:
@brianrodri If I would leave the body empty the linter will yell again:
`Constructor for derived classes must contain a 'super' call`
that's because I extended `ExpressionError` class when I looked at this from the old one:
`ExprUndefinedVarError.protoype = new ExpressionError()`

brianrodri replied:
Ah I didn't notice you were subclassing from `Error` in my first pass. OK, let's try to use `new.target.prototype` instead of hard-coding values: https://stackoverflow.com/a/58417721/4859885

h-ARTS replied:
@brianrodri Great! It's exactly what I was looking for thanks 👍 
Should I add a comment to explain with the link above just like the previous?

brianrodri replied:
Yes please, that would be super helpful (no pun intended)

brianrodri commented on core/templates/expressions/expression-syntax-tree.service.ts:
```diff
@@ -23,54 +23,83 @@ import { AppConstants } from 'app.constants';
+    this.expectedMax = expectedMax;
+  }
+
+  public name: string = 'ExprWrongNumArgsError';
```
Ditto, please remove these.

brianrodri commented on core/templates/expressions/expression-syntax-tree.service.ts:
```diff
@@ -79,32 +108,32 @@ export class ExpressionSyntaxTreeService {
-    };
-
-    let parsed = this.expressionParserService.parse(expression);
+    let uniqueParams = [];
```
Let's just use a `Set` since it removes duplicates for us?

```typescript
let paramsFound = new Set<string>();
paramsFound.add(value);
paramsFound = new Set([...paramsFound, ...this.findParams(parseTree[i]));
return paramsFound;

h-ARTS replied:
@brianrodri I tried with `paramsFound` and I found it a bit hard when declaring `new Set<string>()` and then declaring inside else clause in the for-loop `new Set([...paramsFound, ...this.findParams(parse[i])]);

I tried with the `uniqueParams`:
```
let uniqueParams = new Set<string>();
    for (let i = 0; i < paramsFound.length; i++) {
      if (!uniqueParams.has(paramsFound[i])) {
        uniqueParams.add(paramsFound[i]);
      }
    }
    return Array.from(uniqueParams);
```

brianrodri replied:
I had the following in mind:

```typescript
  public getParamsUsedInExpression(expression: string): string[] {
    let parsedExpression = this.expressionParserService.parse(expression);
    return Array.from(this.findParams(parsedExpression)).sort();
  }

  private findParams(parseTree: string|string[]): Set<string> {
    let paramsFound = new Set<string>();
    if (parseTree instanceof Array) {
      if (parseTree[0] === '#') {
        paramsFound.add(parseTree[1]);
      } else {
        for (let i = 1; i < parseTree.length; i++) {
          paramsFound = new Set<string>([...this.findParams(parseTree[i]), ...paramsFound]);
        }
      }
      return paramsFound;
    }
  }
```

brianrodri replied:
If the unpacking is giving you trouble, we could also do:

```typescript
this.findParams(parseTree[i]).forEach(p => paramsFound.add(p));
```

h-ARTS replied:
@brianrodri I tried that out but the linter tells that you need to activate --downLevelIteration. 
<img width="365" alt="Screenshot 2020-04-01 at 20 09 18" src="https://user-images.githubusercontent.com/11369764/78174177-216e2780-7459-11ea-9a0b-667099613af4.png">

It works after adding it to tsconfig.js but I'm a bit worried if it doesn't add a lot more code after compiling it. I heard it adds helper functions in multiple places which would result in bigger size of code?

Don't have any troubles with unpacking, my end result:
```
public getParamsUsedInExpression(expression: string): string[] {
    let parsed = this.expressionParserService.parse(expression);

    return Array.from(this.findParams(parsed)).sort();
}

private findParams(parseTree: string[]|string): Set<string> {
    let paramsFound = new Set<string>();
    if (parseTree instanceof Array) {
      if (parseTree[0] === '#') {
        paramsFound.add(parseTree[1]);
      } else {
        for (let i = 1; i < parseTree.length; i++) {
          paramsFound = new Set<string>([...this.findParams(parseTree[i]),
            ...paramsFound]);
        }
      }
      return paramsFound;
    }

    let uniqueParams = new Set<string>();
    for (let i = 0; i < paramsFound.size; i++) {
      if (!uniqueParams.has(paramsFound[i])) {
        uniqueParams.add(paramsFound[i]);
      }
    }
    return uniqueParams;
  }
```

brianrodri replied:
I'm confused, on which line is that happening? I don't see any lines that looks like `paramsFound: Set<string>` in the code you pasted? Also, please remove the `uniqueParams` block, `Set` already guarantees unique values which is why I suggested it (you don't have to look for values first).

h-ARTS replied:
This happens on this line of code when spreading `...this.findParams` and `...paramsFound`:
```
paramsFound = new Set<string>([...this.findParams(parseTree[i]), ...paramsFound]);
```

brianrodri replied:
Got it, then let's just do the `forEach` approach:

```typescript
this.findParams(parseTree[i]).forEach(p => paramsFound.add(p));
```

h-ARTS replied:
@brianrodri Ok, I used this approach without `uniqueParams` block and got 1 test failing:
```
HeadlessChrome 80.0.3987 (Mac OS X 10.14.6) Expression evaluator service should get params used in expressions FAILED
	Error: Expected true to be false.
	    at <Jasmine>
	    at core/templates/combined-tests.spec.js:22423:28
	    at <Jasmine>
	    at UserContext.<anonymous> (core/templates/combined-tests.spec.js:22396:11)
	Error: Expected true to be false.
	    at <Jasmine>
	    at core/templates/combined-tests.spec.js:22423:28
	    at <Jasmine>
	    at UserContext.<anonymous> (core/templates/combined-tests.spec.js:22396:11)
	Error: Expected true to be false.
	    at <Jasmine>
	    at core/templates/combined-tests.spec.js:22423:28
	    at <Jasmine>
	    at UserContext.<anonymous> (core/templates/combined-tests.spec.js:22396:11)
	Error: Expected true to be false.
	    at <Jasmine>
	    at core/templates/combined-tests.spec.js:22423:28
	    at <Jasmine>
	    at UserContext.<anonymous> (core/templates/combined-tests.spec.js:22396:11)
	Error: Expected true to be false.
	    at <Jasmine>
	    at core/templates/combined-tests.spec.js:22423:28
	    at <Jasmine>
	    at UserContext.<anonymous> (core/templates/combined-tests.spec.js:22396:11)
	Error: Expected true to be false.
	    at <Jasmine>
	    at core/templates/combined-tests.spec.js:22423:28
	    at <Jasmine>
	    at UserContext.<anonymous> (core/templates/combined-tests.spec.js:22396:11)
	Error: Expected true to be false.
	    at <Jasmine>
	    at core/templates/combined-tests.spec.js:22423:28
	    at <Jasmine>
	    at UserContext.<anonymous> (core/templates/combined-tests.spec.js:22396:11)
	Error: Expected true to be false.
	    at <Jasmine>
	    at core/templates/combined-tests.spec.js:22423:28
	    at <Jasmine>
	    at UserContext.<anonymous> (core/templates/combined-tests.spec.js:22396:11)
	Error: Expected true to be false.
	    at <Jasmine>
	    at core/templates/combined-tests.spec.js:22423:28
	    at <Jasmine>
	    at UserContext.<anonymous> (core/templates/combined-tests.spec.js:22396:11)
	Error: Expected true to be false.
	    at <Jasmine>
	    at core/templates/combined-tests.spec.js:22423:28
	    at <Jasmine>
	    at UserContext.<anonymous> (core/templates/combined-tests.spec.js:22396:11)
	Error: Expected true to be false.
	    at <Jasmine>
	    at core/templates/combined-tests.spec.js:22423:28
	    at <Jasmine>
	    at UserContext.<anonymous> (core/templates/combined-tests.spec.js:22396:11)
	Error: Expected true to be false.
	    at <Jasmine>
	    at core/templates/combined-tests.spec.js:22423:28
	    at <Jasmine>
	    at UserContext.<anonymous> (core/templates/combined-tests.spec.js:22396:11)
	Error: Expected true to be false.
	    at <Jasmine>
	    at core/templates/combined-tests.spec.js:22423:28
	    at <Jasmine>
	    at UserContext.<anonymous> (core/templates/combined-tests.spec.js:22396:11)
```
Doesn't say a lot to me. I tried to console.log the length and items in it but nothing suspecious.

brianrodri replied:
Hmm, are you sure you don't see any extra debug information? I took a look at that code ([link](https://github.com/oppia/oppia/blob/develop/core/templates/expressions/expression-evaluator.service.spec.ts#L61)), and it looks like it tries to print out a lot of info about what failed and why:
```typescript
      var recordFailure = function(params, exception) {
        console.error('input           : ' + expression);
        console.error('parsed          : ' + parsedJson);
        console.error('expected        : ' + JSON.stringify(expectedParams));
        if (params !== undefined) {
          console.error('evaluated       : ' + params);
        } else {
          console.error('exception       : ' + exception);
        }
        failed = true;
      };
```

h-ARTS replied:
@brianrodri Now I understand:
It's not giving the expected parsed result as is should be when looking in `expression-evaluator.service.spec.ts`:
```
HeadlessChrome 80.0.3987 (Mac OS X 10.14.6): Executed 1148 of 1972 SUCCESS (0 secs / 31.221 secs)
ERROR: 'input           : log(9, 3)'
HeadlessChrome 80.0.3987 (Mac OS X 10.14.6): Executed 1148 of 1972 SUCCESS (0 secs / 31.221 secs)
ERROR: 'parsed          : ["log",9,3]'
HeadlessChrome 80.0.3987 (Mac OS X 10.14.6): Executed 1148 of 1972 SUCCESS (0 secs / 31.221 secs)
ERROR: 'expected        : []'
HeadlessChrome 80.0.3987 (Mac OS X 10.14.6): Executed 1148 of 1972 SUCCESS (0 secs / 31.221 secs)
ERROR: 'exception       : TypeError: Cannot read property 'forEach' of undefined'
```

brianrodri replied:
Hmm weird, OK let's do this: Please file an issue to eventually change this function to use `Set`, and assign it to me. Then leave a `TODO(#ISSUE_NUM):` describing the same thing and we can save it for later -- it isn't important enough to block you any more.

h-ARTS replied:
@brianrodri I agree, what about the code itself? Should I wait until the issue gets resolved/merged and then fetch the updated `oppia:develop`?

brianrodri replied:
Nope, I'm fine with merging this as is! Thanks for all the hard work :)

brianrodri commented on core/templates/expressions/expression-syntax-tree.service.ts:
```diff
@@ -201,120 +229,120 @@ export class ExpressionSyntaxTreeService {
         this.verifyNumArgs(args, 2);
-        var numericArgs = this._coerceAllArgsToNumber(args);
+        var numericArgs = this.coerceAllArgsToNumber(args);
         return numericArgs[0] / numericArgs[1];
```
Should we be checking for division by zero here? @seanlip

seanlip replied:
No need for now -- let's keep parity with the existing code. Thanks!

brianrodri commented on core/templates/expressions/expression-syntax-tree.service.ts:
```diff
@@ -79,32 +108,32 @@ export class ExpressionSyntaxTreeService {
 
-    return _findParams(parsed);
-  }
+    return uniqueParams.sort();
```
Let's not sort here in this recursive function, it seems very wasteful. Let's sort the final array that gets returned by `getParamsUsedInExpression` instead:
```typescript
return [...this.findParams(parseTree)].sort();
```

brianrodri commented on core/templates/expressions/expression-syntax-tree.service.ts:
```diff
@@ -23,54 +23,83 @@ import { AppConstants } from 'app.constants';
+}
+
+class ExprUndefinedVarError extends ExpressionError {
+  constructor(public varname: string, public envs: Array<object>) {
```
The body is redundant, please make it empty. Adding "public/private" to the parameters in constructors does the assignments automagically for us.

brianrodri replied:
NVM, didn't notice you were extending `Error`.

brianrodri commented on core/templates/expressions/expression-syntax-tree.service.ts:
```diff
@@ -23,54 +23,83 @@ import { AppConstants } from 'app.constants';
+    this.envs = envs;
+  }
+
+  public name: string = 'ExprUndefinedVarError';
```
`constructor(public name: string)` already does this for us, so this is actually redundant. Please remove.

h-ARTS replied:
@brianrodri That would require me to write in every extended class to write the same which isn't good.
I could do something similar in `ExpressionError`: `this.name = new.target.name` to fix it (I tested) but I'm not sure if that would be a clean solution?

brianrodri replied:
The second sounds fine and clean to me.

brianrodri reviewed (changes requested):
Looking great @h-ARTS, thanks!!

brianrodri commented on core/templates/expressions/expression-syntax-tree.service.ts:
```diff
@@ -16,417 +16,475 @@
+  // throws an error if not. If optional expectedMax is specified, it
+  // verifies the number of args is in [expectedNum, expectedMax] range
+  // inclusive.
+  private verifyNumArgs(args: Array<string>, expectedNum: number,
```
Please use `string[]` instead of `Array<string>` here and elsewhere.

brianrodri commented on core/templates/expressions/expression-syntax-tree.service.ts:
```diff
@@ -16,417 +16,475 @@
 
-      throw new ExprUndefinedVarError(name, envs);
-    };
+  public applyFunctionToParseTree(parsed: string, envs: Array<object>,
```
Similarly, use `any[]` instead of `Array<object>`.

brianrodri commented on core/templates/expressions/expression-syntax-tree.service.ts:
```diff
@@ -16,417 +16,475 @@
+   * can be system or user defined functions and parameters, as well as
+   * system operators.
+   * @param {string} name The name to look up.
+   * @param {!Array.<!object>} envs Represents a nested name space
```
Please remove all type information from the JSDoc's, they are no longer useful since typescript does the type checking for us.

h-ARTS committed:
removed name property and changes to constructor

- constructor left blank
- workaround for restoring prototype chain

h-ARTS committed:
Some minor changes to the types and return statement of getParamsUsedInExpression

brianrodri commented on core/templates/expressions/expression-syntax-tree.service.ts:
```diff
@@ -16,417 +16,475 @@
+    },
+    // Note that removing quotation marks from this key causes issues with
+    // minification (when running the deployment scripts).
+    /* eslint-disable quote-props */
```
Is this comment still relevant? Looks like you're able to remove the quotation marks, so I guess not?

h-ARTS replied:
@brianrodri I didn't notice this. Let me check again.

h-ARTS replied:
👍

h-ARTS committed:
quotation marks added to if, floor, pow, log and abs

h-ARTS committed:
findParams return changed to Set data structure

h-ARTS committed:
Revert "findParams return changed to Set data structure"

This reverts commit 7cef239102bbfa800aae35372a3c7f100f97d9b7.

h-ARTS committed:
TODO Issue added

h-ARTS commented:
@brianrodri I can't assign you: https://github.com/oppia/oppia/issues/8966

brianrodri commented:
> @brianrodri Can't assign you:
> #8966

It's perfect, thanks so much @h-ARTS !!

h-ARTS commented:
@bansalnitish @brianrodri PTAL

brianrodri commented on core/templates/expressions/expression-syntax-tree.service.ts:
```diff
@@ -16,417 +16,444 @@
+    // This code below would restore the protoype chain
+    // Source: https://stackoverflow.com/a/58417721/4859885
+    Object.setPrototypeOf(this, new.target.prototype);
+    this.name = new.target.name;
```
Could you please add a simple test to confirm these work?

h-ARTS replied:
@brianrodri Would this be enough or should I take all the expression error classes and place them in an array?
```
 describe('ExpressionError', () => {
    let expressionError: ExpressionError;

    it('should extend Error object', () => {
      expressionError = new ExpressionError();

      expect(expressionError.name).toBe('ExpressionError');
      expect(expressionError instanceof Error).toBe(true);
    });
  });
```

brianrodri replied:
Just copy and pasting it for each type would be best to keep the tests themselves simple (and would also get you closer to 100% test coverage)

brianrodri reviewed (approved):
Looks beautiful :) Thanks so much, @h-ARTS!

h-ARTS committed:
Tests for expression errors

kevinlee12 commented:
@h-ARTS there's no need to re-request a review if your follow up commit addresses a minor point. However, if your changes are major, then please do so.

kevinlee12 commented:
@brianrodri ptal one more please!

brianrodri reviewed (approved):
Still looks fantastic to me, thanks again for all the hard work @h-ARTS!!

brianrodri commented on core/templates/expressions/expression-syntax-tree.service.spec.ts:
```diff
@@ -15,42 +15,77 @@
+
+    it('should extend ExpressionError class', () => {
+      const exampleVar = undefined;
+      exprUndifinedVarError = new ExprUndefinedVarError(exampleVar, []);
```
Whoops almost missed this, there's a typo here, it should be `exprUndefinedVarError`

brianrodri commented on core/templates/services/UpgradedServices.ts:
```diff
@@ -343,6 +345,8 @@ export class UpgradedServices {
       new ExplorationFeaturesService();
     upgradedServices['ExpressionParserService'] = new ExpressionParserService();
+    upgradedServices['ExpressionSyntaxTreeService'] =
+      new ExpressionSyntaxTreeService(new ExpressionParserService());
```
Do not create a second `ExpressionParserService`. You should pass the one already constructed (on line 347) instead. Also, please move this into the next group of services.

```typescript
upgradedServices['ExpressionSyntaxTreeService'] =
  new ExpressionSyntaxTreeService(upgradedServices['ExpressionParserService']);
```

h-ARTS replied:
> Also, please move this into the next group of services.

@brianrodri Don't get it, the next group of services?

brianrodri replied:
Right, this should be under the `Topological level: 1` group ([link](https://github.com/oppia/oppia/blob/develop/core/templates/services/UpgradedServices.ts#L421)), because it depends on something from the `Topological level: 0` group.

For background, `Topological level: 0` means you take zero arguments (like `ExpressionParserService`).

Level 1 means you take an argument from something in level 0.
Level 2 means you take an argument from something in level 1.
... and so on.

h-ARTS replied:
@brianrodri Ah I see, interesting! 👍

h-ARTS committed:
typo on exprUndefinedVarError

h-ARTS committed:
ExpressionSyntaxTreeService placed in upgraded service in topological level 1

h-ARTS commented:
@bansalnitish PTAL 
Can anyone restart `RUN_E2E_TESTS_CORE_EDITOR_AND_PLAYER_FEATURES=true` in the CI?

kevinlee12 commented:
@h-ARTS I've restarted the test.

abhayraghuwanshi commented:
@h-ARTS could you please take a look here. The test is throwing empty error 
Thanks
````
   it('should throw if environment is not found', () => {
      expect(() => expressionSyntaxTreeService.lookupEnvs('test', [{}]))
        .toThrowError('test not found in [{}]');
    });
````
![Screenshot from 2020-04-05 01-52-04](https://user-images.githubusercontent.com/45534131/78467968-943d0400-76e0-11ea-8385-60bcf3edb9a8.png)
